### ISSUE TITLE
Relational lens tests and unit tests for MySQL and SQLite3

### ISSUE THREAD

**Issue by jamescheney** [links-lang/links#894]
This relies on the changes to the test infrastructure in #893 enabling MySQL and SQLite3 which can be previewed on the `database-ci` branch.

To replicate this for `mysql` do:
```
./run-database-tests tests/relational-lenses -d mysql
```
which results in 
```
!FAILURE: Test 01_cds on database mysql
!FAILURE: Test 02_cds on database mysql
 SUCCESS: Test 03_cds on database mysql
tests/relational-lenses/: 2 failures for mysql
```

~~and for `sqlite3` do:~~
```
./run-database-tests tests/relational-lenses -d sqlite3`
```
~~which results in ~~
```
!FAILURE: Test 01_cds on database sqlite3
!FAILURE: Test 02_cds on database sqlite3
!FAILURE: Test 03_cds on database sqlite3
tests/relational-lenses/: 3 failures for sqlite3
```

To see more details why the tests are failing add `-v` in each case.  From manual inspection it appears something is wrong with the syntax being generated for deletes:
```
***: Fatal error : Mysql returned error: Mysql.exec: You have an error in your SQL syntax; 
check the manual that corresponds to your MySQL server version for the right syntax to 
use near 'DELETE FROM `tracks` WHERE `track` = 'Lovesong' AND `album` = 'Paris';
```

~~For SQLite3 the problems seem to be more fundamental since the assertions are failing with lists of different lengths, seemingly because some of the updates that should be being performed, aren't.~~

**Comment by rudihorn** [comment]
The MySQL bug is most likely due to multiple statement execution. For PHP you do this by passing a flag `CLIENT_MULTI_STATEMENTS` to connect https://www.php.net/manual/en/mysql.constants.php#mysql.client-flags. For us `connect` allows the specification of options, so it might be worth trying to add option `65536` to the connection.

Try changing `mysql_database.ml:245` from
```ocaml
  val connection = connect spec
```
to
```ocaml
  val connection = connect ~options:[65536] spec
```

Perhaps SQLite has similar issues with multiple statements?

**Comment by rudihorn** [comment]
Does the `make_insert_returning_query` work for multiple insertions? This is the one challenge for getting serial columns to work on other databases, as the INSERT RETURNING approach used in postgres isn't supported.

**Comment by jamescheney** [comment]
While it would be good to support sending transactions consisting of multiple updates in one request, I'm not sure what the implications are of this change - is it really something we want to do by default for all databases?  Could we test correctness by issuing each update in a separate request (which might be slower)?  Or is the idea that the updates should be issued atomically in case of concurrent access?  In that case, maybe we should explicitly be using the different APIs for transactions - something that is has also come up in @SimonJF's work on temporal querying.

MySQL and Sqlite3 both have workarounds to support Links's insert returning (which were enabled in #858).  The one caveat is that in MySQL the returned column needs to be declared AUTO_INCREMENT, which is not portable, hence we needed some MySQL-specific setup code for the test that includes insert returning.  Thus, if the tests are also relying on INSERT RETURNING (or if we add tests for the serial column behavior) then maybe the MySQL tests will need to have specific sql initialization code.  I didn't think the basic CD examples made use of serial/autoincrement though.

As far as I am aware the code for insert returning (for all database drivers) is only tested for inserts of a single row.  I'm not sure what the behavior should be if multiple rows are inserted, or whether this is consistent across the different systems.  What is the behavior you rely on (presumably the same as what Postgres does)?  If we do the right thing and use the transaction API, then presumably we can (if necessary) get all of the inserted ids by creating a transaction and issuing multiple insert-returning statements via round trips.

**Comment by rudihorn** [comment]
As far as I know the reason it is disabled by default is that it makes SQL injections more problematic. As Links only allows database access using language integrated query, this shouldn't be much concern to us. This is unrelated to transactions, although it might be worth considering adding them to relational lenses. It is possible to separate the queries into individual queries, but relational lenses are already prone to issuing many queries, and latency can become a big concern here. The issuing of multiple statements in a single query was an optimization necessary to keep performance reasonable, and putting any effort into dismantling this is largely a waste of time if we can just enable it in MySQL.

The tests dont use serial columns I think, and all database drivers need some kind of special declaration to allow auto increment / serial columns. Basic functionality might be portable to all databases. Roundtripping is possible, but slows down bigger changes by a lot and should be avoided if at all possible.

**Comment by rudihorn** [comment]
It seems that SQLite3 may support multiple statements when using `exec` rather than `prepare`.

**Comment by jamescheney** [comment]
Ah, I see - I'll try that and see if it helps.

There's no particular reason for using `prepare` in the SQLite3 driver anyway, as we are always just running the query/statement and parsing the results.  Preparing statements involves a round trip to the database (or in this case SQLite3 C API) to pre-compute the optimized query plan and there's no point in doing this if you are just going to run the query once (or for update statements that you're definitely only going to run once.)

Git blames me for using `prepare`, but I think this driver was originally based on the one for SQLite2 which we no longer support and which I think did the same thing.  Or perhaps when I try to use `prepare` instead of `exec` I'll remember why it had to be `prepare`.

**Comment by jamescheney** [comment]
OK, thanks. 

Sqlite3's ocaml API has a method `prepare_tail` which seems to indicate that `prepare` will only deal with an initial prefix of a statement, but by repeatedly calling `prepare_tail` we can prepare all of them and run them in a single call.  This would be nice since using `exec` would require rewriting (yet again) the code for parsing query results to Links values.  However, the documentation isn't really explicit about how this is supposed to work.

The only issue I can see with multi-statements in MySQL is that if several queries are performed then you potentially get several result sets back and there is a C API for doing this, which ocaml-mysql does not provide access to.  However just enabling multi-statements for sequences of updates (or transactions) with at most one query at the end should be no problem.  We should probably make this expectation explicit in the database interface by offering a method that takes a list of updates and a final (optional) query, submits all of them, and  returns the query result (if any).  (It would be good, for this reason, to distinguish between SQL queries and updates at the type level in sql.ml.)  

This would also accommodate the recently implemented technique for supporting insert-returning in MySQL and SQLite3, which essentially does an ordinary insert and then queries magic rowid fields.  At the moment this is done in a stringly-typed way.

**Comment by jamescheney** [comment]
In ocaml-mysql, the options list is a list of elements of a ML datatype which only gets translated to the flag integers internally, so:
```
  val connection = connect ~options:[65536] spec
```
is not allowed.  Looks like fixing this will require a change to ocaml-mysql - which seems easy enough to patch but who knows when this will be incorporated into a new release of that.

I think for the time being we should have a method `db#exec_multiple : string list * string option -> dbvalue` that expresses the intention to execute multiple updates followed by a query, using a single multi-statement request if possible, but otherwise executes them one at a time, instead of calling `exec` with a string containing multiple statements which will silently (or not so silently) fail with MySQL and SQLite3.

**Comment by rudihorn** [comment]
>I think for the time being we should have a method db#exec_multiple : string list * string option -> dbvalue that expresses the intention to execute multiple updates followed by a query, using a single multi-statement request if possible, but otherwise executes them one at a time, instead of calling exec with a string containing multiple statements which will silently (or not so silently) fail with MySQL and SQLite3.

Hmm this may indeed be necessary. I don't really care about supporting MySQL, but I think SQLite would be valuable to support. I will think about things a bit and see if I can come up with anything else. As far as I can tell though, we could just call `prepare_tail` until it returns `None` and then return the last result, this is how postgresql behaviour anyways.

**Comment by rudihorn** [comment]
The `prepare_tail` variant seems to work for sqlite3. I've pushed the necessary changes to https://github.com/rudihorn/links/tree/database-ci.

**Comment by jamescheney** [comment]
Thanks.  I'd like to get #893 merged first before pulling this in, since it's not urgent; being able to support all three (otherwise) Links-supported databases would be nice before enabling relational lenses by default.  For the time being (pending https://github.com/ygrek/ocaml-mysql/issues/16) we could make MySQL work inefficiently by submitting the updates one at a time, perhaps emitting a warning that this is not a good idea at the moment.

**Comment by rudihorn** [comment]
The branch I sent was based on #893 so I could submit a pull request to merge it into #893 before merging #893?

And as MySQL is definitely not urgend (but will probably be fixed in ocaml mysql) I'd rather not change the code for sql statement generation as it formats an entire batch of arguments in one go for performance rather than concatenating them and if this behaviour can be retained that would be nice.

**Comment by jamescheney** [comment]
The higher-orderness of this makes my head hurt a bit, but I'll give it a try and merge if CI is happy.  I'd prefer to fix MySQL so that it at least works (even if slow) before closing this issue though.  So let's leave this open and see what happens with ocaml-mysql.  The basic change we need is very simple but it may open a can of worms since mulit-statement execution can lead to multi-result handling which also isn't supported by the ocaml-mysql API and would need some API additions.

**Comment by jamescheney** [comment]
Updated title to reflect SQLite3 now fixed in #893, but #893 preserves the existing unit tests for PostgreSQL only which should be generalized to the other two databases once relational lens tests work for all of them.

TODO:

- [ ] Pending fix to ocaml-mysql, test/enable MySQL relational lens tests
- [ ] Generalize unit test framework to run the tests on all three databases in the CI.

**Comment by rudihorn** [comment]
The unit tests could probably do with a fairly large overhaul, as a lot of it predates large lens rewrites and I've also learnt a lot along the way.
